pennylane_lightning/src/tests/Test_GateImplementations_Matrix.cpp

```diff
@@ -0,0 +1,968 @@
+            ComplexPrecisionT{0.2533527906886461, 0.05222114898540775},
+            ComplexPrecisionT{-0.15840033949128557, -0.2727320427534386},
+            ComplexPrecisionT{-0.21590866323269536, -0.1191163626522938},
+        };
+
+        std::vector<ComplexPrecisionT> expected{
+            ComplexPrecisionT{0.140662010561, 0.048572894309},
+            ComplexPrecisionT{0.021854656046, 0.079393113826},
+            ComplexPrecisionT{-0.069686446335, -0.072998481807},
+            ComplexPrecisionT{-0.126819464879, -0.378029601439},
+            ComplexPrecisionT{-0.134716141061, 0.034463497732},
+            ComplexPrecisionT{-0.054908086720, -0.157073288683},
+            ComplexPrecisionT{0.005790250878, -0.346174821950},
+            ComplexPrecisionT{-0.204004872847, 0.019278209615},
+            ComplexPrecisionT{0.336927171579, 0.074028686268},
+            ComplexPrecisionT{0.170822794112, -0.062115684096},
+            ComplexPrecisionT{-0.133087934403, -0.164577625932},
+            ComplexPrecisionT{-0.240412977475, -0.331519081061},
+            ComplexPrecisionT{-0.228436573919, -0.063017646940},
+            ComplexPrecisionT{-0.016556534913, -0.258822480482},
+            ComplexPrecisionT{-0.012416037504, -0.214182329161},
+            ComplexPrecisionT{-0.204751961090, -0.130791666115},
+        };
+
+        auto st = ini_st;
+        GateImplementation::applyMatrix(st.data(), num_qubits, matrix.data(),
+                                        wires, false);
+        REQUIRE(st == PLApprox(expected).margin(1e-5));
+    }
+
+    DYNAMIC_SECTION(GateImplementation::name
+                    << ", Matrix0,1,2,3 - "
+                    << PrecisionToName<PrecisionT>::value) {
+        const size_t num_qubits = 4;
+
+        std::vector<ComplexPrecisionT> ini_st{
+            ComplexPrecisionT{0.203377341216, 0.132238554262},
+            ComplexPrecisionT{0.216290940442, 0.203109511967},
+            ComplexPrecisionT{0.290374372568, 0.123095338906},
+            ComplexPrecisionT{0.040762810130, 0.153237600777},
+            ComplexPrecisionT{0.062445212079, 0.106020046388},
+            ComplexPrecisionT{0.041489260594, 0.149813636657},
+            ComplexPrecisionT{0.002100244854, 0.099744848045},
+            ComplexPrecisionT{0.281559630427, 0.083376695381},
+            ComplexPrecisionT{0.073652349575, 0.066811372960},
+            ComplexPrecisionT{0.150797357980, 0.146266222503},
+            ComplexPrecisionT{0.324043781913, 0.157417591307},
+            ComplexPrecisionT{0.040556496061, 0.254572386140},
+            ComplexPrecisionT{0.204954964152, 0.098550445557},
+            ComplexPrecisionT{0.056681743348, 0.225803880189},
+            ComplexPrecisionT{0.327486634260, 0.130699704247},
+            ComplexPrecisionT{0.299805387808, 0.150417378569},
+        };
+
+        const std::vector<size_t> wires = {0, 1, 2, 3};
+        std::vector<ComplexPrecisionT> matrix{
```

Comment:
This is a big one!

Reply:
Yes, I generated a test example using python. Still, if we are able to use constexpr JSON parser, it would be possible to define all test cases in a JSON file. That would be the best.